File: anna/queue.js
// anna/queue.js
// ─────────────────────────────────────────────────────────────────────────────
// In-process async job queue with bounded concurrency + retry/backoff.
//
// The blueprint's scale model: the bottleneck is API calls, not the language, so
// every Claude / external-integration call runs as a queued job through a worker
// pool. This is a dependency-free in-memory implementation suitable for the
// standalone module and single-process deploys; the same interface (push ->
// promise) can later be backed by a durable queue (e.g. Supabase/pg, Redis)
// without changing callers.
// ─────────────────────────────────────────────────────────────────────────────

/**
 * @param {Object} [opts]
 * @param {number} [opts.concurrency=4]   Max jobs running at once (API rate guard).
 * @param {number} [opts.maxRetries=3]    Retries per job on failure.
 * @param {number} [opts.baseDelayMs=500] Exponential backoff base (500,1000,2000…).
 * @param {function} [opts.onError]       Notified on each failed attempt.
 */
export function createQueue({ concurrency = 4, maxRetries = 3, baseDelayMs = 500, onError } = {}) {
  const pending = [];
  let active = 0;
  const stats = { queued: 0, running: 0, done: 0, failed: 0, retries: 0 };

  const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

  function pump() {
    while (active < concurrency && pending.length) {
      const job = pending.shift();
      active++;
      stats.running = active;
      run(job).finally(() => {
        active--;
        stats.running = active;
        pump();
      });
    }
  }

  async function run(job) {
    let attempt = 0;
    for (;;) {
      try {
        const result = await job.fn();
        stats.done++;
        job.resolve(result);
        return;
      } catch (err) {
        attempt++;
        if (onError) { try { onError(err, { name: job.name, attempt }); } catch { /* ignore */ } }
        if (attempt > maxRetries) {
          stats.failed++;
          job.reject(err);
          return;
        }
        stats.retries++;
        await sleep(baseDelayMs * 2 ** (attempt - 1)); // 500, 1000, 2000, …
      }
    }
  }

  return {
    /**
     * Enqueue a job. `fn` is an async function; resolves/rejects with its result.
     * @returns {Promise<*>}
     */
    push(fn, name = 'job') {
      stats.queued++;
      return new Promise((resolve, reject) => {
        pending.push({ fn, name, resolve, reject });
        pump();
      });
    },
    /** Enqueue many and resolve when all settle. */
    pushAll(fns) { return Promise.all(fns.map((fn, i) => this.push(fn, `batch[${i}]`))); },
    stats: () => ({ ...stats, pending: pending.length }),
    idle: () => active === 0 && pending.length === 0,
  };
}

File: backend/app/anna/queue.py
"""In-process async job queue with bounded concurrency + retry/backoff."""
import asyncio


class _Queue:
    def __init__(self, concurrency=4, max_retries=3, base_delay=0.5, on_error=None):
        self._sem = asyncio.Semaphore(concurrency)
        self.max_retries = max_retries
        self.base_delay = base_delay
        self.on_error = on_error
        self.stats = {"queued": 0, "running": 0, "done": 0, "failed": 0, "retries": 0}

    async def push(self, fn, name="job"):
        self.stats["queued"] += 1
        async with self._sem:
            self.stats["running"] += 1
            try:
                attempt = 0
                while True:
                    try:
                        result = await fn()
                        self.stats["done"] += 1
                        return result
                    except Exception as err:  # noqa: BLE001
                        attempt += 1
                        if self.on_error:
                            try:
                                self.on_error(err, {"name": name, "attempt": attempt})
                            except Exception:  # noqa: BLE001
                                pass
                        if attempt > self.max_retries:
                            self.stats["failed"] += 1
                            raise
                        self.stats["retries"] += 1
                        await asyncio.sleep(self.base_delay * (2 ** (attempt - 1)))
            finally:
                self.stats["running"] -= 1

    async def push_all(self, fns):
        return await asyncio.gather(*[self.push(fn, f"batch[{i}]") for i, fn in enumerate(fns)])


def create_queue(concurrency=4, max_retries=3, base_delay=0.5, on_error=None) -> _Queue:
    return _Queue(concurrency, max_retries, base_delay, on_error)
